### DISCUSSION TITLE
Describe how to package executable with Nuitka

### DISCUSSION THREAD

**Discussion by mrhanCH** [zauberzeug/nicegui#2165] [Ideas / Feature Requests]
The current Pyinstaller packaging method carries the risk of being decompiled, even if key encryption is used, it is useless. Compared to using Nuitka packaging, it is a safer way. Will the Nuitka packaging method be introduced in the future?

**Comment by falkoschindler** [comment]
Hi @mrhanCH!
To be honest, I've never heard of Nuitka packaging before. But it sounds very interesting!
If I understand correctly, we'd basically need another section somewhere around [here](https://nicegui.io/documentation/section_configuration_deployment#package_for_installation) that explains how to package with Nuitka. Is that right? Would you like to create a pull request? 🙂

**Reply by mrhanCH** [reply]
Thank you for your work and sharing！
I am not trying to create a pull request, but rather providing a packaging idea. In my development environment, I encountered a packaging failure using Niutka and received a prompt "FATAL: Error, the Windows SDK must be installed in Visual Studio." Due to some special reasons on my computer, Visual Studio cannot be installed, so it cannot be effectively validated at this time.
I think using Niutka packaging may make the program more secure, at least that's how I feel. I will make time to use other computers, install Visual Studio as required, and then try packaging again.
If there is any new news, I will provide feedback here.

**Reply by xiaokaidada** [reply]
Hopefully pull the request for nuitka I tried packaging with nuitka but it doesn't seem to work，Maybe there's something wrong with my command line

**Comment by mrhanCH** [comment]
> Hopefully pull the request for nuitka I tried packaging with nuitka but it doesn't seem to work，Maybe there's something wrong with my command line

Try my command line, I have tested that it can package and run normally, but the packaging speed is particularly slow.
```bash
nuitka --mingw64 --standalone --show-progress --nofollow-imports --output-dir=out -o test test.py
```

**Reply by retsyo** [reply]
instead of 
```bash
nuitka --mingw64 --standalone --show-progress --nofollow-imports --disable-console --output-dir=out -o test main.py
```
we should use a better solution which 
- does not waste a lot of time
- supplies smaller file size

The steps are
- build `main.pyd` from `main.py` by using cython or nuitka
- make yet another `demo.py` which just import `main.pyd` and run `main` function in it, or just do importation only

the command lines are
- python -m nuitka --mingw64  --show-progress --nofollow-imports   --module --output-dir=out  main.py
- pyinstaller.exe --windowed demo.py

on my windows 10 64bits, python 3.11 64 bits, `NiceGUI\examples\table_and_slots`
- pyinstaller => 94.8M
- nuitka + pyinstaller => 95M
- nuitka => 149M

**Reply by retsyo** [reply]
well I made a pull request https://github.com/zauberzeug/nicegui/pull/2332

**Reply by mrhanCH** [reply]
Yes, there is no problem with your ideas and solutions. 

Initially, I tried to package nuitka on WIN7, but it failed. 

There are actually many packaging optimization solutions, including the one you mentioned that can be faster and smaller. 

Thank you for sharing!

**Comment by xiaokaidada** [comment]
Do you have a test demonstration, I tried but it seems that a lot of environments are missing when packaged
```py
from nicegui import ui

ui.label(text='233')

ui.run(native=True, reload=False)
```

```py
import subprocess

cmd = 'python -m nuitka --mingw64 --standalone --show-progress --nofollow-imports --output-dir=out test.py'

subprocess.call(cmd)
```
Here's my code

**Reply by mrhanCH** [reply]
You need to install mingw64 in advance, save the test code as a test.py file, and then run the above command in your terminal.
If everything goes smoothly, the program will output in the out directory. 
If there is an error during operation, please let me know the error message.
Alternatively, you can try using Nuitka to package other simple testing projects first. If other projects (not Nicegui) cannot be packaged, it may be due to your environment issues.

**Reply by v479038280** [reply]
> Do you have a test demonstration, I tried but it seems that a lot of environments are missing when packaged
> 
> ```python
> from nicegui import ui
> 
> ui.label(text='233')
> 
> ui.run(native=True, reload=False)
> ```
> 
> ```python
> import subprocess
> 
> cmd = 'python -m nuitka --mingw64 --standalone --show-progress --nofollow-imports --output-dir=out test.py'
> 
> subprocess.call(cmd)
> ```
> 
> Here's my code



```
Nuitka-Options:WARNING: Following no imports is unlikely to work for standalone mode and should not be specified.
```
I get this error ,we should remove --standalone.   I test with nuitka 2.3.9
Commercial: None
Python: 3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)]
Flavor: CPython Official
Executable: D:\Program Files\python\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 10
Version C compiler: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\bin\Hostx64\x64\cl.exe (cl 14.3).

**Comment by Aues6uen11Z** [comment]
Additionally, if you do not wish to use `--nofollow-imports`, meaning you want to compile the packages used, you will need to add `--include-package-data=nicegui`.

**Reply by monchin** [reply]
and I found that you need to `--include=module=pygments.formatters.html` as there is a dynamic import

**Comment by laoshaw** [comment]
my nuitka+nicegui single executable build on ubuntu 24.04 is 56MB. nicegui under site-package has 50MB, uvloop is 17MB from fastapi. for a simple fastapi nuitka package it's 27MB.  I don't know how yet, but I believe these sizes can be further trimmed.

**Comment by falkoschindler** [comment]
Reviving this — sorry for the long silence. With #6009 hitting the same wall, we'll add a Nuitka section to the [deployment docs](https://nicegui.io/documentation/section_configuration_deployment#package_for_installation) covering the working invocation (`--standalone` / `--onefile`, `--include-package=nicegui`, `--include-package-data=nicegui`) alongside the existing PyInstaller / `nicegui-pack` instructions. Whether to go further and ship a `nuitka-package.config.yml` so the flags become unnecessary is a separate question we can revisit if there's appetite — for now, documenting what works should unblock people.

**Reply by falkoschindler** [reply]
See PR #6010.
